File: python/categorized_item_factory.py
from item import Item
from categorized_items.normal_item import NormalItem
from categorized_items.backstage_pass import BackstagePass
from categorized_items.legendary_item import LegendaryItem
from categorized_items.well_aging_cheese import WellAgingCheese
from categorized_items.conjured_item import ConjuredItem

SPECIAL_ITEMS = {
    "Backstage passes to a TAFKAL80ETC concert": BackstagePass,
    "Sulfuras, Hand of Ragnaros": LegendaryItem,
    "Aged Brie": WellAgingCheese,
    "Conjured example item": ConjuredItem,
}

class CategorizedItemFactory:
    def _from_item_prefix(item):
        if item.name.startswith("Backstage passes"):
            return BackstagePass(item)
        if item.name.startswith("Conjured"):
            return ConjuredItem(item)
        return None

    def _from_special_item_list(item):
        categorized_item_class = SPECIAL_ITEMS.get(item.name, NormalItem)
        return categorized_item_class(item)

    def from_item(item):
        categorized_item = CategorizedItemFactory._from_item_prefix(item)
        if categorized_item != None:
            return categorized_item

        categorized_item = CategorizedItemFactory._from_special_item_list(item)
        return categorized_item

File: cpp/src/CategorizedItemFactory.cc
#include <string>
#include <vector>

#include "CategorizedItemFactory.h"
#include "categorized_items/NormalItem.cc"
#include "categorized_items/BackstagePass.cc"
#include "categorized_items/AgedCheese.cc"
#include "categorized_items/LegendaryItem.cc"
#include "Inventory.cc"

using namespace std;

Inventory CategorizedItemFactory::inventory = Inventory();

CategorizedItem* CategorizedItemFactory::fromItem(Item &item){
    if (CategorizedItemFactory::itemNameStartsWith(item, "Backstage passes")){
        return new BackstagePass(item);
    }
    if (CategorizedItemFactory::itemNameStartsWith(item, "Conjured")){
        return new NormalItem(item);
    }

    if (inventory.isAgedCheese(item.name)){
        return new AgedCheese(item);
    }
    if (inventory.isLegendary(item.name)){
        return new LegendaryItem(item);
    }

    return new NormalItem(item);
}

bool CategorizedItemFactory::itemNameStartsWith(Item &item, string prefix){
    string s = item.name;
    return s.rfind(prefix, 0) == 0;
}